### PR TITLE
self-signed root cert rotate should also trigger istiod server to reload certificate

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Self-signed root cert rotator does not work when it rotates root cert. It should trigger istiod to resign the certs for its grpc server.

And should update the namespace scoped root cert configmap

### CODE REVIEW

File: pilot/pkg/bootstrap/istio_ca.go [hunk 1]
```diff
@@ -468,6 +467,7 @@ func (s *Server) createIstioCA(opts *caOptions) (*ca.IstioCA, error) {
 
 		s.initCACertsWatcher()
 	}
+	caOpts.OnRootCertUpdate = s.updatePluggedinRootCertAndGenKeyCert
```
Comment (jaellio) [inline pilot/pkg/bootstrap/istio_ca.go:470]: `OnRootCertUpdate` is only used when the cert is self-signed. For clarity, could we move this to the above conditional block for self-signed ca certs?

Comment (jaellio) [inline pilot/pkg/bootstrap/istio_ca.go:470]: The naming of `updatePluggedinRootCertAndGenKeyCert` also seems a bit confusing. The naming makes me think it shouldn't be used when the CA is self-signed and only when it is plugged in by the user. Could it instead be something like `updateDNSCertAndBundle`?

File: security/pkg/pki/ca/ca.go [hunk 2]
```diff
@@ -113,8 +113,13 @@ type IstioCAOptions struct {
 
 	// Config for creating self-signed root cert rotator.
 	RotatorConfig *SelfSignedCARootCertRotatorConfig
+
+	// OnRootCertUpdate is the cb which can only be called by root cert rotator
```
Comment (jaellio) [inline security/pkg/pki/ca/ca.go:117]:
```suggestion
	// OnRootCertUpdate is the cb which can only be called by self-signed root cert rotator
```

File: pilot/pkg/bootstrap/istio_ca.go (hunk 1 above)
Comment (jaellio) [reply pilot/pkg/bootstrap/istio_ca.go:470]: Should we add a debug log line to `updatePluggedinRootCertAndGenKeyCert`?

Comment (jaellio) [inline pilot/pkg/bootstrap/istio_ca.go:470]: Can we capture a full rotation and update of the DNS certificates in an integration test?

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:470]: make sense

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:470]: Yes, will do a little bit improve on this

Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:470]: Not seems very viable, because it can take a very long time.

Comment (jaellio) [reply pilot/pkg/bootstrap/istio_ca.go:470]: Can we modify the root cert validity period?

**Review by linsun — COMMENTED** [review]
Is this only impacting self signed? Given self signed is only used in testing, seems not a big deal to restart istiod during root cert rotate?

**Comment by hzxuzhonghu** [comment]
Yes, only for self signed cert.  I have also thought of removing selfsigned cert rotation as the validity is 10 years by default. 

@howardjohn WDYT

File: pilot/pkg/bootstrap/istio_ca.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/istio_ca.go:470]: minutes? Seems we cannot even accept minutes, because the security test has already exceed 40minutes IIRC

**Comment by hzxuzhonghu** [comment]
@linsun @howardjohn We have two options here:

1. make the rotator work
2. remove it since the default selfsigned root cert is of 10 years validity 

Let me know wdyt please

**Comment by hzxuzhonghu** [comment]
AFAIK, many users make use of selfsigned ca on-premise, so useful for them

**Comment by linsun** [comment]
> AFAIK, many users make use of selfsigned ca on-premise, so useful for them

If we think people are using it on on-prem, I won't object to fix the cert rotate.  I haven't heard people using it other than testing or demo.

**Comment by hzxuzhonghu** [comment]
Then will fix the ci

File: security/pkg/util/certutil.go [hunk 3]
```diff
@@ -55,12 +54,6 @@ func (cu CertUtilImpl) GetWaitTime(certBytes []byte, now time.Time, minGracePeri
 	// Note: multiply time.Duration(int64) by an int (gracePeriodPercentage) will cause overflow (e.g.,
 	// when duration is time.Hour * 90000). So float64 is used instead.
 	gracePeriod := time.Duration(float64(cert.NotAfter.Sub(cert.NotBefore)) * (float64(cu.gracePeriodPercentage) / 100))
-	if gracePeriod < minGracePeriod {
```
Comment (howardjohn) [inline security/pkg/util/certutil.go:58]: why is this removed?

Comment (howardjohn) [reply security/pkg/util/certutil.go:58]: :sleeping: thanks!

**Review by jaellio — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Review by GregHanson — APPROVED** [review]
lgtm
